File: app/src/main/kotlin/jb/openware/app/activity/profile/ProfileEditActivity.java
package in.afi.codekosh.activity.profile;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.net.Uri;
import android.text.TextUtils;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.core.widget.NestedScrollView;

import com.bumptech.glide.Glide;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.GenericTypeIndicator;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Objects;

import de.hdodenhof.circleimageview.CircleImageView;
import in.afi.codekosh.R;
import in.afi.codekosh.tools.BaseFragment;
import in.afi.codekosh.tools.ThemeBuilder;

public class ProfileEditActivity extends BaseFragment {
    private final ArrayList<String> usernames = new ArrayList<>();
    private final FirebaseDatabase _firebase = FirebaseDatabase.getInstance();
    private final DatabaseReference users = _firebase.getReference("Users");
    private String url, name_pre;
    private File avatar;
    private CircleImageView circleImageView1;
    private LinearLayout linearWord;
    private boolean aBoolean;
    private TextInputEditText nameEditText, bioEditText;
    private ImageView back, edit;
    private TextView toolbar, tx_word;
    private View divider;

    @Override
    protected boolean isHomeFragment() {
        return false;
    }

    @Override
    protected int getLayoutRes() {
        return R.layout.activity_profile_edit;
    }

    @Override
    public void getThemeDescriptions(ThemeBuilder themeBuilder) {
        themeBuilder.setImageColorFilter(back, BLACK, WHITE);
        themeBuilder.setImageColorFilter(edit, BLACK, WHITE);

        themeBuilder.setTextColor(toolbar, BLACK, WHITE);

        if (themeBuilder.isNightMode()) {
            divider.setBackgroundColor(ContextCompat.getColor(this, R.color.divider_color_night));
        } else {
            divider.setBackgroundColor(ContextCompat.getColor(this, R.color.divider_color));
        }
    }

    @Override
    protected void initialize() {
        back = findViewById(R.id.back);
        edit = findViewById(R.id.edit);
        toolbar = findViewById(R.id.toolbar);
        tx_word = findViewById(R.id.tx_word);
        Intent intent = getIntent();
        String name = intent.getStringExtra("name");
        url = intent.getStringExtra("url");
        String color = intent.getStringExtra("color");
        String bio_t = intent.getStringExtra("bio");

        avatar = null;
        aBoolean = false;
        nameEditText = findViewById(R.id.name);
        bioEditText = findViewById(R.id.bio);
        circleImageView1 = findViewById(R.id.circleimageview1);
        linearWord = findViewById(R.id.linear_word);


        if (url.equals("none")) {
            hideViews(circleImageView1);
            showViews(linearWord);
            linearWord.setBackground(new GradientDrawable() {
                public GradientDrawable getIns(int a, int b) {
                    this.setCornerRadius(a);
                    this.setColor(b);
                    return this;
                }
            }.getIns(360, Color.parseColor(color)));
            if (name != null) {
                tx_word.setText(name.substring(0, 1));
            }
        } else {
            hideViews(linearWord);
            showViews(circleImageView1);
            Glide.with(this).load(Uri.parse(url)).into(circleImageView1);
        }


        back.setOnClickListener(v -> goBack());
        edit.setOnClickListener(v -> upload());
        findViewById(R.id.picker).setOnClickListener(v -> {
            if (url.equals("none")) {
                pickSinglePhoto((profilePath, imageFileName, imageUri) -> {
                    aBoolean = true;
                    avatar = ImageUtils.compressImage(this, imageUri, 40);
                    hideViews(linearWord);
                    showViews(circleImageView1);
                    Glide.with(this).load(Uri.fromFile(avatar)).into(circleImageView1);
                });
            } else {
                createBottomSheetDialog(R.layout.sheet_profile);
                LinearLayout bt1 = (LinearLayout) bsId(R.id.bt1);
                LinearLayout bt2 = (LinearLayout) bsId(R.id.bt2);
                bt1.setOnClickListener(v1 -> pickSinglePhoto((profilePath, imageFileName, imageUri) -> {
                    aBoolean = true;
                    avatar = ImageUtils.compressImage(this, imageUri, 40);
                    hideViews(linearWord);
                    showViews(circleImageView1);
                    Glide.with(this).load(Uri.fromFile(avatar)).into(circleImageView1);
                    dismissBottomSheetDialog();
                }));
                bt2.setOnClickListener(v1 -> {
                    showProgressDialog();
                    String currentUserId = getUID();
                    DatabaseReference usersDatabaseRef = FirebaseDatabase.getInstance().getReference("Users");

                    ValueEventListener valueEventListener = new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                            if (dataSnapshot.exists() && Objects.equals(dataSnapshot.getKey(), currentUserId)) {
                                HashMap<String, Object> userData = dataSnapshot.getValue(new GenericTypeIndicator<HashMap<String, Object>>() {
                                });
                                if (userData != null) {
                                    String value = String.valueOf(userData.get("avatar_name"));
                                    StorageReference storageRef = FirebaseStorage.getInstance().getReference("avatar").child(value);

                                    storageRef.delete().addOnSuccessListener(unused -> {
                                        HashMap<String, Object> hashMap = new HashMap<>();
                                        hashMap.put("avatar", "none");
                                        hashMap.put("avatar_name", "none");
                                        pushToDatabase(hashMap, "Users", getUID(), unused1 -> {
                                            url = "none";
                                            aBoolean = false;
                                            hideViews(circleImageView1);
                                            showViews(linearWord);
                                            linearWord.setBackground(new GradientDrawable() {
                                                public GradientDrawable getIns(int a, int b) {
                                                    this.setCornerRadius(a);
                                                    this.setColor(b);
                                                    return this;
                                                }
                                            }.getIns(360, Color.parseColor(color)));
                                            if (name != null) {
                                                tx_word.setText(name.substring(0, 1));
                                            }
                                            showToast("Profile picture removed");
                                            dismissProgressDialog();
                                        }, e -> {
                                            dismissProgressDialog();
                                            alertCreator(e.getMessage());
                                        });

                                    }).addOnFailureListener(e -> alertCreator(e.getMessage()));
                                }
                            }
                        }

                        @Override
                        public void onCancelled(@NonNull DatabaseError databaseError) {
                            // Handle onCancelled event, if needed
                            dismissProgressDialog();
                            alertCreator("An error has occurred. Please try again later.");
                        }
                    };
                    usersDatabaseRef.child(currentUserId).addListenerForSingleValueEvent(valueEventListener);
                    dismissBottomSheetDialog();
                });
                showBottomSheetDialog();

            }
        });
        nameEditText.setText(name);
        bioEditText.setText(bio_t);
    }

    @Override
    protected void initializeLogic() {
        setUpDivider();
        back.setOnClickListener(v -> goBack());
        ChildEventListener _users_child_listener = new ChildEventListener() {
            @Override
            public void onChildAdded(@NonNull DataSnapshot _param1, String _param2) {
                GenericTypeIndicator<HashMap<String, Object>> _ind = new GenericTypeIndicator<HashMap<String, Object>>() {
                };
                final String _childKey = _param1.getKey();
                final HashMap<String, Object> _childValue = _param1.getValue(_ind);
                if (_childValue != null && _childValue.containsKey("name")) {
                    usernames.add((String) _childValue.get("name"));
                }
                if (_childKey != null && _childKey.equals(getUID()) && _childValue != null) {
                    name_pre = String.valueOf(_childValue.get("name"));
                }

            }

            @Override
            public void onChildChanged(@NonNull DataSnapshot _param1, String _param2) {
                GenericTypeIndicator<HashMap<String, Object>> _ind = new GenericTypeIndicator<HashMap<String, Object>>() {
                };
                final String _childKey = _param1.getKey();
                final HashMap<String, Object> _childValue = _param1.getValue(_ind);
                if (_childValue != null && _childValue.containsKey("name")) {
                    usernames.add((String) _childValue.get("name"));
                }
                if (_childKey != null && _childKey.equals(getUID()) && _childValue != null) {
                    name_pre = String.valueOf(_childValue.get("name"));
                }

            }

            @Override
            public void onChildMoved(@NonNull DataSnapshot _param1, String _param2) {

            }

            @Override
            public void onChildRemoved(@NonNull DataSnapshot _param1) {

            }

            @Override
            public void onCancelled(@NonNull DatabaseError _param1) {

            }
        };
        users.addChildEventListener(_users_child_listener);
    }

    private void setUpDivider() {
        NestedScrollView nestedScrollView = findViewById(R.id.nested_scroll_view);
        divider = findViewById(R.id.divider);
        divider.setVisibility(View.GONE);
        nestedScrollView.setVerticalScrollBarEnabled(false);
        nestedScrollView.setOnScrollChangeListener((NestedScrollView.OnScrollChangeListener) (v, scrollX, scrollY, oldScrollX, oldScrollY) -> divider.setVisibility(scrollY > 0 ? View.VISIBLE : View.GONE));
    }


    private void upload() {
        closeKeyboard();
        showProgressDialog();
        String name = String.valueOf(nameEditText.getText());
        String bio = String.valueOf(bioEditText.getText());
        String currentUserId = getUID();
        DatabaseReference usersDatabaseRef = FirebaseDatabase.getInstance().getReference("Users");
        if (TextUtils.isEmpty(name)) {
            showToast("Name empty");
            dismissProgressDialog();
        } else if (TextUtils.isEmpty(bio)) {
            showToast("Bio is empty");
            dismissProgressDialog();
        } else if (!name_pre.equals(name) && usernames.contains(name)) {
            new MaterialAlertDialogBuilder(this).setTitle("Alert").setMessage("Username already taken.").setPositiveButton("Ok", null).show();
            dismissProgressDialog();
        } else {
            if (aBoolean) {
                if (!url.equals("none")) {

                    ValueEventListener valueEventListener = new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                            if (dataSnapshot.exists() && Objects.equals(dataSnapshot.getKey(), currentUserId)) {
                                HashMap<String, Object> userData = dataSnapshot.getValue(new GenericTypeIndicator<HashMap<String, Object>>() {
                                });
                                if (userData != null) {
                                    String value = Objects.requireNonNull(userData.get("avatar_name")).toString();
                                    StorageReference storageRef = FirebaseStorage.getInstance().getReference("avatar").child(value);

                                    // Delete the file from Firebase Storage
                                    storageRef.delete().addOnSuccessListener(unused -> push()).addOnFailureListener(e -> push());
                                }
                            }
                        }

                        @Override
                        public void onCancelled(@NonNull DatabaseError databaseError) {
                            // Handle onCancelled event, if needed
                            alertCreator("An error has occurred. Please try again later.");
                        }
                    };
                    usersDatabaseRef.child(currentUserId).addListenerForSingleValueEvent(valueEventListener);
                } else {
                    push();
                }
            } else {
                HashMap<String, Object> dataMap = new HashMap<>();
                dataMap.put("name", String.valueOf(nameEditText.getText()));
                dataMap.put("bio", String.valueOf(bioEditText.getText()));
                String databasePath = "Users";
                String child = getUserConfig().getUid();
                pushToDatabase(dataMap, databasePath, child, unused -> {
                    dismissProgressDialog();
                    finish();
                }, e -> alertCreator(e.getMessage()));
            }
        }
    }

    private void push() {
        String s = generateFileNameWithTimestamp();
        uploadFileToFirebaseStorage("avatar", s, Uri.fromFile(avatar), uri -> {
            String fileDownloadUrl = uri.toString();
            HashMap<String, Object> dataMap = new HashMap<>();
            dataMap.put("avatar", fileDownloadUrl);
            dataMap.put("name", String.valueOf(nameEditText.getText()));
            dataMap.put("avatar_name", s);
            String databasePath = "Users";
            String child = getUserConfig().getUid();
            pushToDatabase(dataMap, databasePath, child, unused -> {
                dismissProgressDialog();
                finish();
            }, e -> alertCreator(e.getMessage()));
        }, e -> alertCreator(e.getMessage()));
    }

}

File: app/src/main/kotlin/jb/openware/app/ui/activity/profile/ProfileEditActivity.kt
package jb.openware.app.ui.activity.profile

import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.widget.LinearLayout
import androidx.core.graphics.toColorInt
import androidx.core.net.toUri
import com.bumptech.glide.Glide
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.firebase.database.ChildEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.storage.FirebaseStorage
import jb.openware.app.R
import jb.openware.app.databinding.ActivityProfileEditBinding
import jb.openware.app.ui.common.BaseActivity
import jb.openware.app.ui.components.BottomSheetController
import jb.openware.app.ui.items.UserProfile
import jb.openware.app.util.ImageUtil
import jb.openware.app.util.Utils
import java.io.File

class ProfileEditActivity :
    BaseActivity<ActivityProfileEditBinding>(ActivityProfileEditBinding::inflate) {

    private val usersRef = FirebaseDatabase.getInstance().getReference("Users")

    private val usernames = mutableSetOf<String>()
    private var url: String = "none"
    private var namePre: String = ""
    private var avatar: File? = null
    private var pickedNewAvatar = false

    private var name: String? = null
    private lateinit var color: String

    override fun init() {
        name = intent.getStringExtra("name")
        color = intent.getStringExtra("color") ?: "#000000"
        val bioText = intent.getStringExtra("bio") ?: ""
        url = intent.getStringExtra("url") ?: "none"

        pickedNewAvatar = false
        avatar = null

        // avatar display
        if (url == "none") {
            hideViews(binding.circleimageview)
            showViews(binding.linearWord)
            binding.linearWord.background = createRoundedDrawable(color.toColorInt())
            name?.let { if (it.isNotEmpty()) binding.txWord.text = it.first().toString() }
        } else {
            showViews(binding.circleimageview)
            hideViews(binding.linearWord)
            Glide.with(this).load(url.toUri()).into(binding.circleimageview)
        }

        binding.toolbar.setNavigationOnClickListener(Utils.getBackPressedClickListener(this))

        binding.toolbar.setOnMenuItemClickListener { item ->
            if (item.itemId == R.id.action_upload) {
                upload()
                true
            } else false
        }

        binding.name.setText(name)
        binding.bio.setText(bioText)

    }

    override fun initLogic() {
        observeUsers()
        setupAvatarPicker()
    }

    private fun observeUsers() {
        usersRef.addChildEventListener(object : ChildEventListener {
            override fun onChildAdded(s: DataSnapshot, p: String?) = handleUser(s)
            override fun onChildChanged(s: DataSnapshot, p: String?) = handleUser(s)
            override fun onChildMoved(s: DataSnapshot, p: String?) {}
            override fun onChildRemoved(s: DataSnapshot) {}
            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun handleUser(snapshot: DataSnapshot) {
        val user = snapshot.getValue(UserProfile::class.java) ?: return

        user.name.let { usernames.add(it) }

        if (snapshot.key == getUid()) namePre = user.name
    }

    private fun setupAvatarPicker() {
        binding.picker.setOnClickListener {

            // PICK NEW
            if (url == "none") {
                pickSinglePhoto { _, _, uri ->
                    pickedNewAvatar = true
                    avatar = ImageUtil.compressImage(this, uri, 40)
                    showAvatarPreview()
                }
                return@setOnClickListener
            }

            // REMOVE / CHANGE OPTIONS
            val sheet = BottomSheetController(this).apply { create(R.layout.sheet_profile) }
            sheet.find<LinearLayout>(R.id.bt1).setOnClickListener { // Change
                pickSinglePhoto { _, _, uri ->
                    pickedNewAvatar = true
                    avatar = ImageUtil.compressImage(this, uri, 40)
                    showAvatarPreview()
                    sheet.dismiss()
                }
            }

            sheet.find<LinearLayout>(R.id.bt2).setOnClickListener { // Remove
                removeAvatar(sheet)
            }

            sheet.show()
        }
    }

    private fun showAvatarPreview() {
        hideViews(binding.linearWord)
        showViews(binding.circleimageview)
        Glide.with(this).load(Uri.fromFile(avatar)).into(binding.circleimageview)
    }

    private fun removeAvatar(sheet: BottomSheetController) {
        showProgressDialog()
        val id = getUid()

        usersRef.child(id).get().addOnSuccessListener { snap ->
            val user = snap.getValue(UserProfile::class.java)
            val avatarUrl = user?.avatar ?: ""

            if (avatarUrl.isNotBlank() && avatarUrl != "none") {
                FirebaseStorage.getInstance().getReferenceFromUrl(avatarUrl).delete()
                    .addOnCompleteListener { updateAvatarNone(sheet) }
            } else {
                updateAvatarNone(sheet)
            }
        }.addOnFailureListener {
            dismissProgressDialog()
            alertCreator("Failed to remove avatar")
        }
    }

    private fun updateAvatarNone(sheet: BottomSheetController) {
        val map = mapOf("avatar" to "none")
        pushToDatabase(map, "Users", getUid(), {
            url = "none"
            pickedNewAvatar = false
            hideViews(binding.circleimageview)
            showViews(binding.linearWord)
            binding.linearWord.background = createRoundedDrawable(color.toColorInt())
            binding.txWord.text = namePre.firstOrNull()?.toString() ?: ""
            toast("Profile picture removed")
            dismissProgressDialog()
            sheet.dismiss()
        }) {
            dismissProgressDialog()
            alertCreator(it.message)
        }
    }

    private fun upload() {
        hideKeyboard()
        showProgressDialog()

        val inputName = binding.name.text.toString()
        val inputBio = binding.bio.text.toString()
        val id = getUid()

        when {
            inputName.isBlank() -> return fail("Name empty")
            inputBio.isBlank() -> return fail("Bio is empty")
            namePre != inputName && usernames.contains(inputName) -> {
                MaterialAlertDialogBuilder(this).setTitle("Alert")
                    .setMessage("Username already taken.").setPositiveButton("Ok", null).show()
                return fail()
            }
        }

        // if new avatar selected
        if (pickedNewAvatar) {
            handleAvatarUpload(inputName, inputBio)
            return
        }

        // normal update
        val data = mapOf("name" to inputName, "bio" to inputBio)
        pushToDatabase(data, "Users", id, {
            dismissProgressDialog()
            finish()
        }) { fail(it.message) }
    }

    private fun handleAvatarUpload(name: String, bio: String) {
        if (url != "none") removeOldAvatarThenPush(name, bio) else pushNewAvatar(name, bio)
    }

    private fun removeOldAvatarThenPush(name: String, bio: String) {
        val id = getUid()
        usersRef.child(id).get().addOnSuccessListener { snap ->
            val avatarUrl = snap.getValue(UserProfile::class.java)?.avatar ?: ""

            if (avatarUrl.isNotBlank() && avatarUrl != "none") {
                FirebaseStorage.getInstance().getReferenceFromUrl(avatarUrl).delete()
                    .addOnCompleteListener { pushNewAvatar(name, bio) }
            } else pushNewAvatar(name, bio)
        }.addOnFailureListener { pushNewAvatar(name, bio) }
    }

    private fun pushNewAvatar(name: String, bio: String) {
        val file = avatar ?: return fail("No avatar file found")
        val fileName = generateFileNameWithTimestamp()
        uploadFileToFirebaseStorage("avatar", fileName, file.toUri(), { uri ->
            val map = mapOf(
                "avatar" to uri.toString(), "name" to name, "bio" to bio
            )
            pushToDatabase(map, "Users", getUid(), {
                dismissProgressDialog()
                finish()
            }) { fail(it.message) }
        }) { fail(it.message) }
    }

    private fun fail(msg: String? = null) {
        msg?.let { toast(it) }
        dismissProgressDialog()
    }

    private fun createRoundedDrawable(color: Int) = GradientDrawable().apply {
        cornerRadius = 360f
        setColor(color)
    }

}